File: legacy-node/src/localityKey.js
/**
 * Hierarchical locality keys: {ISO3166-1}:{ISO3166-2}:{postal}
 * Examples: IN:TN:600115, IN:TN, IN
 */

const LOCALITY_KEY_PART = /^[A-Z0-9]{2,10}$/;

export function normalizeLocalityKey(key) {
  const trimmed = String(key ?? "").trim();
  if (!trimmed) {
    return "";
  }
  return trimmed
    .split(":")
    .map((part) => part.trim().toUpperCase())
    .filter(Boolean)
    .join(":");
}

/**
 * @param {string} key
 * @returns {boolean}
 */
export function isValidLocalityKey(key) {
  const normalized = normalizeLocalityKey(key);
  if (!normalized) {
    return false;
  }
  const parts = normalized.split(":");
  if (parts.length < 1 || parts.length > 3) {
    return false;
  }
  return parts.every((part) => LOCALITY_KEY_PART.test(part));
}

/**
 * Most-specific first: IN:TN:600115 → IN:TN → IN
 * @param {string} key
 * @returns {string[]}
 */
export function localityKeyChain(key) {
  const trimmed = String(key ?? "").trim();
  if (!trimmed) {
    return [];
  }
  const parts = trimmed.split(":");
  const chain = [];
  for (let depth = parts.length; depth >= 1; depth -= 1) {
    chain.push(parts.slice(0, depth).join(":"));
  }
  return chain;
}

/**
 * True when `offerKey` applies to demand at `userKey` (equal or ancestor prefix).
 * @param {string} offerKey
 * @param {string} userKey
 */
export function offerAppliesToLocality(offerKey, userKey) {
  const offer = String(offerKey ?? "").trim();
  const user = String(userKey ?? "").trim();
  if (!offer || !user) {
    return false;
  }
  if (offer === user) {
    return true;
  }
  return user.startsWith(`${offer}:`);
}

/**
 * True when a record's locality key equals the filter or is a descendant
 * (deeper segment chain). Examples with record IN:TN:600097:
 *   filter IN        → match
 *   filter IN:TN     → match
 *   filter IN:TN:600097 → match
 *   filter IN:KA     → no match
 * @param {string} recordKey
 * @param {string} filterKey
 */
export function recordMatchesLocalityFilter(recordKey, filterKey) {
  const record = normalizeLocalityKey(recordKey);
  const filter = normalizeLocalityKey(filterKey);
  if (!record || !filter) {
    return false;
  }
  if (record === filter) {
    return true;
  }
  return record.startsWith(`${filter}:`);
}

/**
 * Pick catalog rows for a resolved user locality; most-specific menu line wins per offer id.
 * @param {Array<{ id?: string, standard_offer_id?: string, locality_key: string, menu_label?: string }>} offers
 * @param {string} userLocalityKey
 */
export function resolveStandardOffersForLocality(offers, userLocalityKey) {
  const userKey = String(userLocalityKey ?? "").trim();
  if (!userKey || !Array.isArray(offers)) {
    return [];
  }
  const chain = new Set(localityKeyChain(userKey));
  const applicable = offers.filter((offer) =>
    chain.has(String(offer.locality_key ?? "").trim())
  );
  const byOfferId = new Map();
  for (const offer of applicable) {
    const offerId = String(offer.id ?? offer.standard_offer_id ?? "").trim();
    if (!offerId) {
      continue;
    }
    const depth = String(offer.locality_key ?? "").split(":").length;
    const existing = byOfferId.get(offerId);
    if (!existing || depth > existing.depth) {
      byOfferId.set(offerId, { offer, depth });
    }
  }
  return [...byOfferId.values()]
    .map((entry) => entry.offer)
    .sort((a, b) =>
      String(a.menu_label ?? "").localeCompare(String(b.menu_label ?? ""))
    );
}

File: src/main/java/org/sharingbridge/integration/geo/LocalityKey.java
package org.sharingbridge.integration.geo;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

/**
 * Hierarchical locality keys: {@code {ISO3166-1}:{ISO3166-2}:{postal}}.
 * Examples: {@code IN:TN:600115}, {@code IN:TN}, {@code IN}. Matches Node {@code localityKey.js}.
 */
public final class LocalityKey {

    private static final Pattern LOCALITY_KEY_PART = Pattern.compile("^[A-Z0-9]{2,10}$");

    private LocalityKey() {}

    public static String normalizeLocalityKey(Object key) {
        String trimmed = jsString(key).trim();
        if (trimmed.isEmpty()) {
            return "";
        }
        String[] parts = trimmed.split(":");
        List<String> normalized = new ArrayList<>();
        for (String part : parts) {
            String p = part.trim().toUpperCase(Locale.ROOT);
            if (!p.isEmpty()) {
                normalized.add(p);
            }
        }
        return String.join(":", normalized);
    }

    public static boolean isValidLocalityKey(Object key) {
        String normalized = normalizeLocalityKey(key);
        if (normalized.isEmpty()) {
            return false;
        }
        String[] parts = normalized.split(":");
        if (parts.length < 1 || parts.length > 3) {
            return false;
        }
        for (String part : parts) {
            if (!LOCALITY_KEY_PART.matcher(part).matches()) {
                return false;
            }
        }
        return true;
    }

    /** Most-specific first: IN:TN:600115 → IN:TN → IN */
    public static List<String> localityKeyChain(Object key) {
        String trimmed = jsString(key).trim();
        if (trimmed.isEmpty()) {
            return List.of();
        }
        String[] parts = trimmed.split(":");
        List<String> chain = new ArrayList<>();
        for (int depth = parts.length; depth >= 1; depth--) {
            chain.add(String.join(":", java.util.Arrays.copyOf(parts, depth)));
        }
        return chain;
    }

    /**
     * True when {@code offerKey} applies to demand at {@code userKey} (equal or ancestor prefix).
     */
    public static boolean offerAppliesToLocality(Object offerKey, Object userKey) {
        String offer = jsString(offerKey).trim();
        String user = jsString(userKey).trim();
        if (offer.isEmpty() || user.isEmpty()) {
            return false;
        }
        if (offer.equals(user)) {
            return true;
        }
        return user.startsWith(offer + ":");
    }

    /**
     * True when a record's locality key equals the filter or is a descendant (deeper segment chain).
     */
    public static boolean recordMatchesLocalityFilter(Object recordKey, Object filterKey) {
        String record = normalizeLocalityKey(recordKey);
        String filter = normalizeLocalityKey(filterKey);
        if (record.isEmpty() || filter.isEmpty()) {
            return false;
        }
        if (record.equals(filter)) {
            return true;
        }
        return record.startsWith(filter + ":");
    }

    /**
     * Pick catalog rows for a resolved user locality; most-specific menu line wins per offer id.
     */
    public static List<Map<String, Object>> resolveStandardOffersForLocality(
            List<? extends Map<String, ?>> offers, Object userLocalityKey) {
        String userKey = jsString(userLocalityKey).trim();
        if (userKey.isEmpty() || offers == null) {
            return List.of();
        }
        Set<String> chain = Set.copyOf(localityKeyChain(userKey));
        List<Map<String, Object>> applicable = new ArrayList<>();
        for (Map<String, ?> offer : offers) {
            if (offer == null) {
                continue;
            }
            Object lk = offer.get("locality_key");
            String locality = jsString(lk).trim();
            if (chain.contains(locality)) {
                applicable.add(new LinkedHashMap<>(offer));
            }
        }
        Map<String, OfferDepth> byOfferId = new LinkedHashMap<>();
        for (Map<String, Object> offer : applicable) {
            Object idObj = offer.get("id");
            if (idObj == null) {
                idObj = offer.get("standard_offer_id");
            }
            String offerId = jsString(idObj).trim();
            if (offerId.isEmpty()) {
                continue;
            }
            Object lk = offer.get("locality_key");
            int depth = jsString(lk).split(":").length;
            OfferDepth existing = byOfferId.get(offerId);
            if (existing == null || depth > existing.depth) {
                byOfferId.put(offerId, new OfferDepth(offer, depth));
            }
        }
        return byOfferId.values().stream()
                .map(entry -> entry.offer)
                .sorted(Comparator.comparing(
                        o -> String.valueOf(o.getOrDefault("menu_label", "")),
                        String::compareTo))
                .collect(Collectors.toList());
    }

    /** Matches JS {@code String(value ?? "")}. */
    private static String jsString(Object value) {
        return value == null ? "" : String.valueOf(value);
    }

    private record OfferDepth(Map<String, Object> offer, int depth) {}
}
